pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go

```diff
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package reconcilerv2
+
+import (
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/cilium/cilium/pkg/bgpv1/manager/store"
+	"github.com/cilium/cilium/pkg/bgpv1/types"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	"github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/labels"
+	slim_metav1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
+)
+
+type (
+	// PeerAdvertisements is a map of peer name to its family advertisements
+	// This is the top level map that is returned to the consumer with requested advertisements.
+	PeerAdvertisements       map[string]PeerFamilyAdvertisements
+	PeerFamilyAdvertisements map[v2alpha1.CiliumBGPFamily][]v2alpha1.BGPAdvertisement // key is the address family type
+)
+
+type PeerAdvertisementIn struct {
+	cell.In
+
+	Logger          logrus.FieldLogger
+	PeerConfigStore store.BGPCPResourceStore[*v2alpha1.CiliumBGPPeerConfig]
+	AdvertStore     store.BGPCPResourceStore[*v2alpha1.CiliumBGPAdvertisement]
+}
+
+type CiliumPeerAdvertisement struct {
+	logger     logrus.FieldLogger
+	peerConfig store.BGPCPResourceStore[*v2alpha1.CiliumBGPPeerConfig]
+	adverts    store.BGPCPResourceStore[*v2alpha1.CiliumBGPAdvertisement]
+}
+
+func NewCiliumPeerAdvertisement(p PeerAdvertisementIn) *CiliumPeerAdvertisement {
+	return &CiliumPeerAdvertisement{
+		logger:     p.Logger,
+		peerConfig: p.PeerConfigStore,
+		adverts:    p.AdvertStore,
+	}
+}
+
+// GetConfiguredAdvertisements can be called to get all configured advertisements of given BGPAdvertisementType for each peer.
+// Advertisements are selected based on below criteria:
+// Each peer is selected from the BGP node instance configuration. For each peer, the peer configuration is fetched
+// from local store.
+// Peer configuration contains the list of families and the advertisement selector.
+// We iterate over all advertisements ( available from local store ), select only those that match the advertisement
+// selector of the family.
+// Information of peer -> family -> advertisements is returned to the consumer.
+func (p *CiliumPeerAdvertisement) GetConfiguredAdvertisements(conf *v2alpha1.CiliumBGPNodeInstance, selectAdvertTypes ...v2alpha1.BGPAdvertisementType) (PeerAdvertisements, error) {
```

Comment:
I think it is bit early to optimize. The main factor will be number of peers. Number of advertisements are pretty much constant.

 Linear scan [ Peers ] - O(n) ( number of peers )
 Linear scan [ Families ] - O(m) ( max 2 )
 Linear scan [ Advertisements ] - O(k) ( number of advertisements - 3-4 types, which is again filtered)

Overall complexity will be factor of number of peers. Even if we cache `type + families -> advertisements`, we will need to do linear scan per peer to get correct advertisement. Since number of families and advertisement configurations are expected to be small - I am not very concerned with this approach. 

We can def revisit this if we see it becoming a bottlenech - need some metrics :)